[18.0] [ADD] l10n_it_edi_extension

Borruso opened the pull request.

SirAionTech committed:
[IMP] l10n_it_edi_extension: Create partner when not found

Borruso committed:
[ADD] l10n_it_edi_extension: permit to have preview of e-invoice

Borruso committed:
[ADD] l10n_it_edi_extension: adding tags in xml

Borruso committed:
[ADD] l10n_it_edi_extension: permit import e-invoice details

eLBati commented:
Non mi è chiaro: quali sono le funzionalità di questo modulo? Non si sovrappongono a quelle di `l10n_it_edi`?

francesco-ooops commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,12 @@
+**Italiano**
```
Dato che è stata fatta un'analisi approfondita, per qualsiasi nuovo modulo per la 18 ne approfitterei per inserire una documentazione più approfondita, non necessariamente step-by-step ma sicuramente più ampia di così

Borruso replied:
sicuramente piano piano andremo ad aggiornare più che la descrizione la configurazione/funzionamento (CONFIG.md/USAGE.md)

francesco-ooops replied:
Benissimo, sarebbe una cosa buona da fare prima del merge (magari con l'aiuto di un funzionale), altrimenti resta sempre una di quelle cose da fare dopo 👍

eLBati commented on l10n_it_edi_extension/models/ir_attachment.py:
```diff
@@ -0,0 +1,62 @@
+        return "FoglioStileAssoSoftware.xsl"
+
+    @api.model
+    def remove_xades_sign(self, xml_string):
```
Non ci sono metodi di `l10n_it_edi` per fare questo? A `l10n_it_edi` non serve farlo?

Borruso replied:
controllo ma forse non l'avevo trovato

eLBati commented on l10n_it_edi_extension/views/account_move_view.xml:
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Tutti questi dati li vogliamo anche nel form della fattura? Non è sufficiente leggerli eventualmente dall'XML?

Borruso replied:
sono gli stessi campi che abbiamo in v16 quando estrapoliamo le info dall'xml

eLBati replied:
Lo so :-)

eLBati commented on l10n_it_edi_extension/views/l10n_it_view.xml:
```diff
@@ -0,0 +1,40 @@
+                    <div class="o_row" invisible="not l10n_it_edi_attachment_id">
+                        <field
+                            name="l10n_it_edi_attachment_id"
+                            groups="account.group_account_invoice"
```
qua l'obiettivo è cambiare il gruppo di `l10n_it_edi_attachment_id`? In tal caso non è meglio farlo direttamente con `attribute` invece che metterlo invisibile e riaggiungerlo dopo?

Borruso replied:
in realtà non è solo per il gruppo è per aggiungere il pulsante per la preview

eLBati commented on l10n_it_edi_extension/wizards/edi_import_file_wizard.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class EdiImportFileWizard(models.TransientModel):
+    _name = "edi.import.file.wizard"
```
Se è per l'importazione di uno ZIP con tanti XML, lo chiamerei di conseguenza

Borruso replied:
suggerimenti?

eLBati replied:
`edi.import.zip.file.wizard`

eLBati commented:
I campi introdotti da questo modulo hanno bisogno di script di upgrade da v16, per recuperare dati che erano in campi gestiti da `l10n_it_fatturapa*`.
Abbiamo però il problema che, durante l'upgrade alla 18, `l10n_it_edi` non è installabile per via dell'`excludes` di `l10n_it_fatturapa`.
Questo vincolo scrive nella tabella `ir_module_module_exclusion`.
Bisognerebbe quindi provare a rimuovere il vincolo in `ir_module_module_exclusion` in uno script di `_pre_init_hook` di `l10n_it_edi_extension`.

Poi possibilmente, una volta che tutti i dati sono stati portati, disinstallare `l10n_it_fatturapa` (teoricamente in `_post_init_hook`)

Borruso committed:
[ADD] l10n_it_edi_extension: adding pre/post init hook

monen17 commented on l10n_it_edi_extension/__init__.py:
```diff
@@ -0,0 +1,1188 @@
+from odoo.addons.base.models.ir_qweb_fields import Markup, nl2br, nl2br_enclose
+
+OLD_MODULES = [
+    "l10n_it_fatturapa",
+    "l10n_it_fatturapa_in",
+    "l10n_it_fatturapa_out",
+    "l10n_it_fatturapa_sale",
+    "l10n_it_fiscal_payment_term",
+    "l10n_it_fiscalcode",
+    "l10n_it_ipa",
+    "l10n_it_pec",
+    "l10n_it_rea",
+    "l10n_it_vat_payability",
```
Un gran lavoro! :clap:
Purtroppo non posso provarlo perché OpenUpgrade ancora non ha migrato gran che alla `18.0`.

Se questo però è il modulo che dovrebbe coprire le funzionalità di questi moduli, mi aspetto che ci siano dei test.
Non dico tutti quelli dei moduli elencati, che potrebbero facilmente essere un centinaio, però abbiamo fatto un grande lavoro sui test nella `16.0` e non mi va che vada tutto perso.

Borruso replied:
per i test in realtà fa tutto `l10n_it_edi` posso fare solo test per preview dell'xml e la popolazione delle tabella di dettaglio della fattura elettronica

monen17 replied:
Ho preso il primo test di uno dei moduli elencati: https://github.com/OCA/l10n-italy/blob/56621cb667a6d9ed581a55bd452508b28d49e3da/l10n_it_fiscalcode/tests/test_fiscalcode.py#L17-L34

`l10n_it_edi` come fa il calcolo del codice fiscale di un partner?

Per sapere se tutte le funzionalità dei nostri vecchi moduli ci sono in `l10n_it_edi` c'è un modo semplice: eseguire i test automatici.

monen17 reviewed (changes requested):
Grazie della PR!
Oltre a quanto scritto sotto, concordo con le altre revisioni che vada quantomeno esteso il README.

Borruso committed:
[ADD] l10n_it_edi_extension: manage compute fiscal code on partner

monen17 commented:
Ho proposto https://github.com/Borruso/l10n-italy/pull/23 per creare il partner se mancante durante l'import.

odooNextev committed:
[IMP] l10n_it_edi_extension: add readme

monen17 committed:
[IMP] l10n_it_edi_extension: Import multiple invoices

monen17 commented:
Ho proposto https://github.com/Borruso/l10n-italy/pull/25 per creare tutte le fatture quando viene importata una fattura elettronica contenente più fatture.

andreampiovesana commented on l10n_it_edi_extension/README.rst:
```diff
@@ -46,6 +117,81 @@ This module adds base feature to handle e-invoice.
+1. Siano installati e configurati i moduli dipendenti:
+
+-  account: Il modulo base della contabilità di Odoo deve essere
+   installato e configurato (piano dei conti, tasse, diari contabili,
```
meglio giornali contabili (non diari)

monen17 committed:
[IMP] l10n_it_edi_extension: Import detail level

Allow the user to import e-bills with no lines, one line for each tax rate, or all lines (default).

eLBati commented on l10n_it_edi_extension/wizards/compute_fc.py:
```diff
@@ -0,0 +1,214 @@
+
+
+class WizardComputeFc(models.TransientModel):
+    _name = "wizard.compute.fc"
```
Questo non è documentato nel `readme` mi pare

odooNextev replied:
https://github.com/Borruso/l10n-italy/pull/29

eLBati commented on l10n_it_edi_extension/wizards/l10n_it_edi_import_file_wizard.py:
```diff
@@ -0,0 +1,98 @@
+
+
+class EInvoiceImportFileWizard(models.TransientModel):
+    _name = "l10n_it_edi.import_file_wizard"
```
Questo non è documentato nel `readme` mi pare

odooNextev replied:
https://github.com/Borruso/l10n-italy/pull/29

eLBati commented on l10n_it_edi_extension/data/invoice_it_template.xml:
```diff
@@ -0,0 +1,52 @@
+        </xpath>
+        <xpath expr="//DatiGeneraliDocumento/ImportoTotaleDocumento" position="after">
+            <t t-foreach="causale" t-as="c">
+                <Causale t-out="c" />
```
Anche questo è da documentare nel README

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/28.

matteotognini commented:
Ho testato la preview dell'xml e funziona correttamente.

monen17 commented:
Ho proposto https://github.com/Borruso/l10n-italy/pull/27 per importare una fattura elettronica senza righe, con una riga per ogni aliquota, oppure con tutte le righe (default).

odooNextev committed:
[IMP] l10n_it_edi_extension: improve readme

odooNextev commented:
@Borruso non capisco perchè falliscono i test: ho provato in locale e non riscontro l'errore della divisione per 0 come qui https://github.com/OCA/l10n-italy/actions/runs/15343445807/job/43174440067?pr=4411#step:8:321
Ne ottengo altri 2 nel file `test_import_zip` che qui non ci sono, ma contabilmente funziona:

- uno per validazione errata di questo CF: MRORSS90E25B111T
- uno perchè `skipping IT05979361218_003.xml.p7m, not an XML/P7M file` e quindi la lunghezza di in_moves non è più 36, ma 35

micheledic commented:
> @Borruso non capisco perchè falliscono i test: ho provato in locale e non riscontro l'errore della divisione per 0 come qui https://github.com/OCA/l10n-italy/actions/runs/15343445807/job/43174440067?pr=4411#step:8:321 Ne ottengo altri 2 nel file `test_import_zip` che qui non ci sono, ma contabilmente funziona:
> 
> * uno per validazione errata di questo CF: MRORSS90E25B111T
> * uno perchè `skipping IT05979361218_003.xml.p7m, not an XML/P7M file` e quindi la lunghezza di in_moves non è più 36, ma 35

Probabilmente in locale non hai il core di odoo pullato.
Il problema esce perchè prima c'era una errata gestione degli sconti maggiorazione  che ho fixato con https://github.com/odoo/odoo/commit/d2b1b140b95f6c60d3a032745ee6880688fa2748 .
Non è questa mia PR a rompere, ma il file presente nello zip di test 

> subdir1/IT01234567890_FPR04.xml

che ha prezzo e sconto a 0 .
```
 <Quantita>1.00</Quantita>
        <UnitaMisura>Nr</UnitaMisura>
        <PrezzoUnitario>0.00</PrezzoUnitario>
        <ScontoMaggiorazione>
          <Tipo>SC</Tipo>
          <Percentuale>0.00</Percentuale>
          <Importo>0.00</Importo>
        </ScontoMaggiorazione>
        <PrezzoTotale>0.00</PrezzoTotale>
        <AliquotaIVA>22.00</AliquotaIVA>
```
Imho non ha senso avere lo sconto a 0 e su una riga a 0 (fermo restando che provvederò comunque a migliorare quella funzione con altra pr)
@Borruso @SirAionTech

odooNextev commented:
Grazie @micheledic per la spiegazione, effettivamente non avevo il core aggiornato

Borruso commented:
@eLBati i test vanno i errore perché prezzo unitario è zero e Odoo non fa il check prima di dividere per prezzo unitario
```
File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1361, in _l10n_it_edi_import_line
    discount = 100 - (100 * current_unit_price) / move_line.price_unit
```

Ho trovato la fattura nei test dello zip IT01234567890_FPR04.xml (non so se è unica)
![FPR04](https://github.com/user-attachments/assets/023986b3-32f1-4f7f-a67f-9fa1fb237264)

tafaRU commented on l10n_it_edi_extension/__init__.py:
```diff
@@ -0,0 +1,1188 @@
+]
+
+
+def rename_fields(env, table, field_updates, condition=None):
```
come mai non usi dappertutto [rename_fields](https://github.com/OCA/openupgradelib/blob/d2d76e772760365988a53af11b65eab648d991d2/openupgradelib/openupgrade.py#L675) di `openupgralib`?

tafaRU replied:
@Borruso mi sono accorto che anche [qui](https://github.com/OCA/l10n-italy/blob/2de18b1a726fb4887eda79d8d5fae1e10e50807b/l10n_it_account/migrations/18.0.1.0.0/post-migration.py#L12) non hai usato il metodo messo a disposizione da `openupgradelib`. Attendo tua risposta prima di aprire evntualmente una issue grazie.

eLBati commented:
@Borruso @micheledic 

Il file `IT01234567890_FPR04.xml` è valido per SDI 

![image](https://github.com/user-attachments/assets/9109a021-828c-4215-8ce4-fa4461007eee)

quindi odoo lo dovrebbe importare senza errori, ma:

```
2025-07-04 13:05:37,094 1 ERROR test_import_e_bill odoo.addons.account.models.account_move: Errore durante l'importazione dell'allegato 'IT01234567890_FPR04.xml' come fattura (decodificatore=_l10n_it_edi_import_invoice) 
Traceback (most recent call last):
  File "/opt/odoo/addons/account/models/account_move.py", line 4187, in _extend_with_attachments
    success = decoder(invoice, file_data, new)
  File "/opt/odoo/addons/l10n_it_edi_ndd/models/account_move.py", line 91, in _l10n_it_edi_import_invoice
    res = super()._l10n_it_edi_import_invoice(invoice=invoice, data=data, is_new=is_new)
  File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1205, in _l10n_it_edi_import_invoice
    message_to_log += self._l10n_it_edi_import_line(element, move_line, extra_info)
  File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1361, in _l10n_it_edi_import_line
    discount = 100 - (100 * current_unit_price) / move_line.price_unit
ZeroDivisionError: float division by zero
```

Quindi bisogna segnalarlo a odoo, se non già fatto

Borruso commented:
> @Borruso @micheledic
> 
> Il file `IT01234567890_FPR04.xml` è valido per SDI
> 
> ![image](https://private-user-images.githubusercontent.com/1033131/462497145-9109a021-828c-4215-8ce4-fa4461007eee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE2MzUwNjAsIm5iZiI6MTc1MTYzNDc2MCwicGF0aCI6Ii8xMDMzMTMxLzQ2MjQ5NzE0NS05MTA5YTAyMS04MjhjLTQyMTUtOGNlNC1mYTQ0NjEwMDdlZWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDRUMTMxMjQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5MWQ3NDY0MTkxODQxNjY4OTVhZmM4ZDAxOTczMDNmY2E0MTZiOWMzM2QzMzBmODVkMGNmMDdhMmUzMDI1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xbV485jMFK5hh767IysRRyOnKgzw-lyvZ0ftnQbD4Gc)
> 
> quindi odoo lo dovrebbe importare senza errori, ma:
> 
> ```
> 2025-07-04 13:05:37,094 1 ERROR test_import_e_bill odoo.addons.account.models.account_move: Errore durante l'importazione dell'allegato 'IT01234567890_FPR04.xml' come fattura (decodificatore=_l10n_it_edi_import_invoice) 
> Traceback (most recent call last):
>   File "/opt/odoo/addons/account/models/account_move.py", line 4187, in _extend_with_attachments
>     success = decoder(invoice, file_data, new)
>   File "/opt/odoo/addons/l10n_it_edi_ndd/models/account_move.py", line 91, in _l10n_it_edi_import_invoice
>     res = super()._l10n_it_edi_import_invoice(invoice=invoice, data=data, is_new=is_new)
>   File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1205, in _l10n_it_edi_import_invoice
>     message_to_log += self._l10n_it_edi_import_line(element, move_line, extra_info)
>   File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1361, in _l10n_it_edi_import_line
>     discount = 100 - (100 * current_unit_price) / move_line.price_unit
> ZeroDivisionError: float division by zero
> ```
> 
> Quindi bisogna segnalarlo a odoo, se non già fatto

yes

micheledic commented:
> > @Borruso @micheledic
> > Il file `IT01234567890_FPR04.xml` è valido per SDI
> > ![image](https://private-user-images.githubusercontent.com/1033131/462497145-9109a021-828c-4215-8ce4-fa4461007eee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE2MzUwNjAsIm5iZiI6MTc1MTYzNDc2MCwicGF0aCI6Ii8xMDMzMTMxLzQ2MjQ5NzE0NS05MTA5YTAyMS04MjhjLTQyMTUtOGNlNC1mYTQ0NjEwMDdlZWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDRUMTMxMjQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5MWQ3NDY0MTkxODQxNjY4OTVhZmM4ZDAxOTczMDNmY2E0MTZiOWMzM2QzMzBmODVkMGNmMDdhMmUzMDI1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xbV485jMFK5hh767IysRRyOnKgzw-lyvZ0ftnQbD4Gc)
> > quindi odoo lo dovrebbe importare senza errori, ma:
> > ```
> > 2025-07-04 13:05:37,094 1 ERROR test_import_e_bill odoo.addons.account.models.account_move: Errore durante l'importazione dell'allegato 'IT01234567890_FPR04.xml' come fattura (decodificatore=_l10n_it_edi_import_invoice) 
> > Traceback (most recent call last):
> >   File "/opt/odoo/addons/account/models/account_move.py", line 4187, in _extend_with_attachments
> >     success = decoder(invoice, file_data, new)
> >   File "/opt/odoo/addons/l10n_it_edi_ndd/models/account_move.py", line 91, in _l10n_it_edi_import_invoice
> >     res = super()._l10n_it_edi_import_invoice(invoice=invoice, data=data, is_new=is_new)
> >   File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1205, in _l10n_it_edi_import_invoice
> >     message_to_log += self._l10n_it_edi_import_line(element, move_line, extra_info)
> >   File "/opt/odoo/addons/l10n_it_edi/models/account_move.py", line 1361, in _l10n_it_edi_import_line
> >     discount = 100 - (100 * current_unit_price) / move_line.price_unit
> > ZeroDivisionError: float division by zero
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Quindi bisogna segnalarlo a odoo, se non già fatto
> 
> yes

https://github.com/odoo/odoo/pull/217670 attendiamo merge

odooNextev committed:
[FIX] l10n_it_edi_extension: fix migration script for fc, rea and rd

eLBati commented on l10n_it_edi_extension/__manifest__.py:
```diff
@@ -0,0 +1,36 @@
+    "name": "Italy - E-invoicing - Base Feature",
+    "version": "18.0.1.0.0",
+    "category": "Accounting/Localizations/EDI",
+    "development_status": "Production/Stable",
```
Meglio Beta

matteotognini commented:
Ho testato la creazione automatica dei partner all'import della fattura, funziona.

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+1. Anteprima e Download del file XML:
+
+   - Aggiunge un pulsante ("Preview XML") direttamente nel form della fattura.
+  Questo pulsante permette di visualizzare un'anteprima del file XML della fattura elettronica prima dell'invio effettivo.
```
Come si fa a vedere l'anteprima del file prima dell'invio?
Ho fatto qualche prova e il pulsante per vedere l'anteprima è visibile solo quando la fattura elettronica viene creata, e viene creato all'invio.

odooNextev replied:
hai ragione, correggo

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+   - Aggiunge un pulsante ("Preview XML") direttamente nel form della fattura.
+  Questo pulsante permette di visualizzare un'anteprima del file XML della fattura elettronica prima dell'invio effettivo.
+   - Dalla stessa finestra di anteprima, è possibile scaricare il file XML generato.
+   - Aggiunta della possibilità di visualizzare l'anteprima del file XML anche dal portale.
```
Come si fa a vedere l'anteprima nel portale? Non ho visto nulla nel codice o nelle viste che lo permetta.

odooNextev replied:
nel portale effettivamente non c'è l'anteprima, si può solo scaricare il file XML insieme al PDF, ma penso sia di default di `l10n_it_edi`

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+   - Dalla stessa finestra di anteprima, è possibile scaricare il file XML generato.
+   - Aggiunta della possibilità di visualizzare l'anteprima del file XML anche dal portale.
+
+2. Aggiunge campi nell'export delle fatture XML:
```
Non credo si capisca molto quello che effettivamente vorrebbe dire, forse qualcosa del genere sarebbe più chiaro?
```suggestion
2. Valorizza i seguenti campi/nodi della fattura elettronica esportata:
```
Sarebbe anche utile scrivere (qui o in CONFIGURE) dove sono i campi in Odoo da valorizzare per avere questi nodi nel XML: alcuni vanno messi nella fattura, altri nella company, altri nelle righe.

monen17 replied:
Manca dove andrebbero impostati i campi

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+2. Aggiunge campi nell'export delle fatture XML:
+
+   - `<RiferimentoAmministrazione>` (sia sulla riga che nei dati generali): identificativo utilizzato per uso amministrativo/gestionale interno. È un campo libero che può essere utilizzato per inserire riferimenti specifici richiesti dalla Pubblica Amministrazione o altri riferimenti utili per la gestione amministrativa.
```
Come si fa ad aggiungere nelle righe? Non trovo il campo nelle viste.

eLBati replied:
C'è da aggiungere il campo (optional) nella vista delle righe fattura

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/data/invoice_it_template.xml:
```diff
@@ -0,0 +1,52 @@
+                </Indirizzo>
+                <CAP>
+                    <t t-out="partner.zip" />
+                </CAP>
```
Puoi mettere le direttive `t-*` direttamente dentro il nodo? Altrimenti vengono aggiunti spazi prima/dopo il valore: 
```xml
<StabileOrganizzazione>
  <Indirizzo> str </Indirizzo>
  <CAP> 12345 </CAP>
  <Comune>city</Comune>
  <Provincia>RM</Provincia>
  <Nazione>IT</Nazione>
</StabileOrganizzazione>
```

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/data/invoice_it_template.xml:
```diff
@@ -0,0 +1,52 @@
+            <t t-foreach="causale" t-as="c">
+                <Causale t-out="c" />
+            </t>
+            <Art73 t-if="company.l10n_edi_it_art73">SI</Art73>
```
Questo è un campo del documento, quindi della fattura, perché è legato invece alla `company`?

eLBati replied:
Nella 16 era sull'azienda, lo possiamo lasciare sull'azienda

eLBati replied:
In realtà @Borruso lo ha già aggiunto anche su account.move

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+   - Aggiunge un'opzione per creare i contatti presenti in una fattura elettronica se non esistono in anagrafica tra cui:
+     - `<CessionarioCommittente>` e `<CedentePrestatore>` invece di scrivere solamente i dettagli nel chatter.
+     - `<RappresentanteFiscale>`
+     - `<TerzoIntermediarioOSoggettoEmittente>`
```
A quanto vedo questi vengono creati sempre se mancano, quindi la loro creazione non dipende da nessuna opzione.

eLBati replied:
Anche nella 16 vengono creati sempre, quindi lo possiamo tenere così e modificare il README

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                    vat,
+                    codice_fiscale,
+                ):
+                    self.l10n_it_edi_intermediary_id = intermediary.id
```
Tutto questo è duplicato (a parte le stringhe e i nomi delle variabili), puoi estrarlo in un metodo? Si potrebbe chiamare `..search_create..`.

monen17 replied:
Ho visto il refactoring grazie, ho proposto altre modifiche in https://github.com/Borruso/l10n-italy/pull/33.

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            else:
+                if tax_representative := self._l10n_it_edi_create_partner(
+                    body_tree, "//RappresentanteFiscale", vat, codice_fiscale
+                ):
```
`_l10n_it_edi_search_partner` già farebbe la creazione del partner quando non lo trova, quindi perché:
1. Aggiungere nel contesto `skip_create` così `search_partner` non creerà il partner
2. chiamare `search_partner` che non creerà il partner
3. se il partner non è stato trovato, chiamare `create_partner`

?

eLBati replied:
Nel contesto invece dello skip mettiamo quale nodo dell'XML leggere per eventualmente creare il partner

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            ):
+                self.l10n_it_edi_tax_representative_id = tax_representative.id
+            else:
+                if tax_representative := self._l10n_it_edi_create_partner(
```
Anche a me piace usare `:=`, ma `_l10n_it_edi_create_partner` creerà sempre un partner, quindi a cosa serve l'`if`?

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                )
+                message_to_log.append(message)
+
+        if elements_parent_invoice := body_tree.xpath(
+            ".//DatiGenerali/FatturaPrincipale"
+        ):
+            for element_parent_invoice in elements_parent_invoice:
+                self.write(
+                    {
+                        "l10n_it_edi_related_invoice_code": get_text(
+                            element_parent_invoice, ".//NumeroFatturaPrincipale"
+                        ),
+                        "l10n_it_edi_related_invoice_date": get_date(
+                            element_parent_invoice, ".//DataFatturaPrincipale"
+                        ),
+                    }
+                )
```
C'è solo un nodo `FatturaPrincipale`, perché quindi un `for` loop? Tra l'altro se ci fossero più nodi i valori dei campi assegnati verrebbero sovrascritti.

monen17 replied:
La correzione è in https://github.com/Borruso/l10n-italy/pull/33.

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+
+        if elements_summary := body_tree.xpath(".//DatiBeniServizi/DatiRiepilogo"):
+            for element_summary in elements_summary:
+                self.env["l10n_it_edi.summary_data"].create(
```
Possiamo usare invece la creazione in batch? È fatta apposta per ottimizzare queste situazioni.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+
+        if elements_sal := body_tree.xpath(".//DatiGenerali/DatiSAL"):
+            for element_sal in elements_sal:
+                self.env["l10n_it_edi.activity_progress"].create(
```
Possiamo usare invece la creazione in batch? È fatta apposta per ottimizzare queste situazioni.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+        ]:
+            if body_tree.xpath(xpath):
+                message = Markup("<br/>").join(
+                    (self.env._(label), self._compose_info_message(body_tree, xpath))
```
```suggestion
            (".//DatiGenerali/DatiTrasporto", self.env._("Transport informations from XML file:")),
            (".//DatiVeicoli", self.env._("Vehicle informations from XML file:")),
        ]:
            if body_tree.xpath(xpath):
                message = Markup("<br/>").join(
                    label, self._compose_info_message(body_tree, xpath))
```
Altrimenti queste stringhe non vengono esportate come da tradurre perché `_` non è abbastanza furbo per capire cosa c'è dentro `label`

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+
+            if elements_code := element.xpath(".//CodiceArticolo"):
+                for element_code in elements_code:
+                    self.env["l10n_it_edi.article_code"].create(
```
Possiamo usare invece la creazione in batch? È fatta apposta per ottimizzare queste situazioni.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+
+            if elements_discount := element.xpath(".//ScontoMaggiorazione"):
+                for element_discount in elements_discount:
+                    self.env["l10n_it_edi.discount_rise_price"].create(
```
Possiamo usare invece la creazione in batch? È fatta apposta per ottimizzare queste situazioni.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+
+            if elements_other_data := element.xpath(".//AltriDatiGestionali"):
+                for element_other_data in elements_other_data:
+                    self.env["l10n_it_edi.line_other_data"].create(
```
Possiamo usare invece la creazione in batch? È fatta apposta per ottimizzare queste situazioni.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                    "Import detail level %(import_detail_level)s not supported.\n"
+                    "Please set an import detail level in company %(company)s.",
+                    import_detail_level=import_detail_level,
+                    company=company,
```
```suggestion
                    company=company.name,
```

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+       Nel file XML, questi codici sono nel nodo `<CodiceArticolo>` che può contenere:
+       - `<CodiceTipo>`: identifica il tipo di codifica utilizzata
+       - `<CodiceValore>`: il valore effettivo del codice
+     - Miglioramento della gestione dello sconto o maggiorazione, nodo `<ScontoMaggiorazione>`
```
Che tipo di miglioramento? Vedo che crea dei record

eLBati replied:
Semplicamente registriamo `ScontoMaggiorazione` nelle righe di dettagli della fattura importata. Possiamo modificare il README in questo senso

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+     - senza righe
+     - una riga per ogni aliquota
+     - tutte le righe (default)
+   - Gestione avanzata dei `<DatiRiepilogo>` con l'importazione di:
```
```suggestion
   - Importazione dei `<DatiRiepilogo>`:
```
perché mi pare che crei solo dei record all'importazione.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+     - `<Imposta>`
+     - `<EsigibilitaIVA>`: Indica quando l'IVA diventa esigibile (immediata, differita o scissione dei pagamenti)
+     - `<RiferimentoNormativo>`: Obbligatorio quando si usa il campo `<Natura>`
+   - Gestione avanzata dei `<DatiGeneraliDocumento>` con l'importazione di:
```
```suggestion
   - Importazione dei `<DatiGeneraliDocumento>`:
```
perché mi pare che crei solo dei record all'importazione.

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/wizards/compute_fc.py:
```diff
@@ -0,0 +1,214 @@
+
+except ImportError:
+    _logger.warning(
+        "codicefiscale library not found. "
```
È una dipendenza del modulo, perché dovrebbe non esserci?

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+      - controllo sul totale IVA inclusa
+      - miglioramento della validazione del codice fiscale
+
+4. Aggiunge di un wizard per calcolare i codici fiscali attingendo dai dati dei comuni italiani reperibili dal sito dell'AdE http://www.agenziaentrate.gov.it/wps/content/Nsilib/Nsi/Strumenti/Codici+attivita+e+tributo/Codici+territorio/Comuni+italia+esteri
```
A me questo link dà Access Denied, ma poi davvero viene chiamato nel modulo? Non trovo dove.
E non mi pare lo faccia nemmeno la libreria `codicefiscale` usata.

eLBati replied:
Non funziona più e non sappiamo se c'è un link aggiornato, quindi lo possiamo togliere per ora

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/wizards/compute_fc.py:
```diff
@@ -0,0 +1,214 @@
+
+        it = self.env.ref("base.it").id
+        res = {
+            "domain": {"birth_province": [("country_id", "=", it)]},
```
Ritornare un `domain` veniva ammonito con un warning in `16.0`, adesso viene proprio ignorato, vedi https://github.com/odoo/odoo/commit/024fadfd0edf94147a58ffe11c719baafa73dd23.
Si può togliere?

Se serve un dominio dinamico va implementato nel `domain` dei campi.

eLBati replied:
`domain` da togliere

eLBati replied:
per @monen17  da qua in giù

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+   - `<RiferimentoAmministrazione>` (both on line and in general data): identifier used for internal administrative/management purposes. It's a free field that can be used to insert specific references required by Public Administration or other useful references for administrative management.
+   - `<StabileOrganizzazione>`: represents the data of the seller/provider's permanent establishment in Italy if different from the registered office
+   - `<Causale>`: in this case there is no specific field, but it transcribes the "Terms and conditions" od the invoice.
```
```suggestion
   - `<Causale>`: in this case there is no specific field, but it transcribes the "Terms and conditions" of the invoice.
```

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+4. Aggiunge di un wizard per calcolare i codici fiscali attingendo dai dati dei comuni italiani reperibili dal sito dell'AdE http://www.agenziaentrate.gov.it/wps/content/Nsilib/Nsi/Strumenti/Codici+attivita+e+tributo/Codici+territorio/Comuni+italia+esteri
+
+5. Importazione di una fattura elettronica senza righe, con una riga per ogni aliquota, oppure con tutte le righe (default).
```
Questo c'è già più su:
>    - Aggiunge la possibilità di scegliere tra 3 modalità di importazione:
     - senza righe
     - una riga per ogni aliquota
     - tutte le righe (default)

eLBati replied:
Togliamo questo e lasciamo quello sopra

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+5. Importazione di una fattura elettronica senza righe, con una riga per ogni aliquota, oppure con tutte le righe (default).
+
+6. Il campo "Termini e condizioni" viene scritto nel nodo "Causale" della fattura elettronica.
```
Questo c'è già più su:
> - `<Causale>`: in questo caso non c'è un campo apposito, ma trascrive i "Termini e condizioni" della fattura

eLBati replied:
da togliere

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,168 @@
+
+**English**
+
+his module extends Odoo's standard Italian electronic invoicing functionality, introducing useful tools such as XML preview and fiscal code calculation and improvements in importing and exporting XML invoices.
```
```suggestion
This module extends Odoo's standard Italian electronic invoicing functionality, introducing useful tools such as XML preview and fiscal code calculation and improvements in importing and exporting XML invoices.
```

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            and not partner
+            and not self.env.context.get("skip_create_partner")
+        ):
+            try:
```
Ho provato a importare una fattura da interfaccia con il pulsante "Upload" ma non si entra qui dentro, quando dovrebbe essere chiamata questa parte di codice?

eLBati replied:
Spostiamo l'eventuale creazione dei partner in `_l10n_it_edi_import_invoice` con `_l10n_it_edi_extension_create_partner` , dicendogli quale nodo usare

monen17 replied:
@Borruso di questo abbiamo parlato in chiamata e avevo capito ci stessi già lavorando perché è citato anche sopra, fai tu giusto?

Borruso replied:
sisi fatto

monen17 replied:
Come mai è rimasta questa parte?
Avevo capito che sarebbe stata da spostare, quindi rimuovere da qui e mettere da un'altra parte.

eLBati replied:
L'override di `_l10n_it_edi_search_partner` è da togliere.

La parte relativa a stabile organizzazione la mettiamo in `_l10n_it_edi_get_extra_info`

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
Non si parla dell'aggiornamento dei partner durante l'importazione: è configurabile tramite un flag nel partner stesso.

eLBati replied:
da aggiungere

monen17 replied:
@Borruso fammi sapere se questa feature rimarrà dopo lo spostamento della creazione del partner.

@eLBati se non rimane, non è più da aggiungere.

Borruso replied:
non ho capito a quale ti riferisci

monen17 replied:
> non ho capito a quale ti riferisci

L'aggiornamento dei partner durante l'importazione che è configurabile tramite un flag nel partner stesso.

Borruso replied:
è rimasto

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            return datetime.fromisoformat(datetime_str.replace("Z", "+00:00"))
+        except (ValueError, TypeError):
+            return False
+    return False
```
Tutte queste funzioni sono già implementate in https://github.com/odoo/odoo/blob/b7fd61a70a7f41c863529afde493a723a114d5be/addons/l10n_it_edi/models/account_move.py#L22-L50 e mi pare siano identiche, perché non importiamo quelle invece di averne qui una copia?

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+        "will create a line in the bill.",
+    )
+
+    @api.constrains("l10n_it_codice_fiscale", "company_type")
+    def validate_codice_fiscale(self):
+        res = super().validate_codice_fiscale()
+        for partner in self:
+            if not partner.l10n_it_codice_fiscale:
+                continue
+            elif (
+                partner.company_type == "person"
+                and len(partner.l10n_it_codice_fiscale) != 16
+            ):
+                msg = self.env._("The fiscal code must have 16 characters.")
+                raise ValidationError(msg)
+        return res
```
La validazione del codice fiscale in `16.0` è più avanzata di così:
https://github.com/OCA/l10n-italy/blob/11cb1bd550881ea23fa23d830c6dbf00ccb5a0d7/l10n_it_fiscalcode/model/res_partner.py#L13-L40
si può migrare quella?

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
Non si parla dei controlli sugli importi che producono inconsistenze e fanno visualizzare i messaggi sulla fattura.

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move_line.py:
```diff
@@ -0,0 +1,16 @@
+    _inherit = "account.move.line"
+
+    l10n_it_edi_admin_ref = fields.Char(string="Admin. ref.", size=20, copy=False)
+    l10n_it_edi_discount_rise_price_ids = fields.One2many(
```
Come mai questi due campi sono su `account.move.line` mentre tutti gli altri campi della fattura elettronica sono su `l10n_it_edi.line`?
Tra l'altro in `l10n_it_edi.line` ci sono anche campi equivalenti a questi:
```py
    l10n_it_edi_discount_rise_price_ids = fields.One2many(
        "l10n_it_edi.discount_rise_price",
        "l10n_it_edi_line_id",
        string="Discount and Supplement Details",
        readonly=True,
    )
    admin_ref = fields.Char(string="Administration Reference", readonly=True)
```

eLBati replied:
Visto che lo abbiamo su `l10n_it_edi.line`, possiamo toglierlo da `account.move.line`

eLBati replied:
`l10n_it_edi_admin_ref`, visto che lo abbiamo già su `account.move.line` (e il campo ce lo teniamo perchè lo usiamo in esportazione), lo possiamo togliere da `l10n_it_edi.line`

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move_line.py:
```diff
@@ -0,0 +1,16 @@
+    _inherit = "account.move.line"
+
+    l10n_it_edi_admin_ref = fields.Char(string="Admin. ref.", size=20, copy=False)
+    l10n_it_edi_discount_rise_price_ids = fields.One2many(
```
Esiste già il campo `discount` per lo sconto su `account.move.line`, che rapporto c'è con questi record?

eLBati replied:
Vedi sopra

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            countries = self.env["res.country"].search([("code", "=", country_code)])
+            if countries:
+                country_id = fields.first(countries).id
+            else:
```
Ce ne sarà probabilmente solo una, ma perché cercarle tutte per prendere solo la prima?
```suggestion
            country_id = self.env["res.country"].search([("code", "=", country_code)], limit=1).id
            if not country_id:
```

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                )
+
+        vals = {
+            "vat": country_code + vat,
```
Qui chiederei un parere di @TheMule71 che lo so che gli piace l'argomento :wink:

eLBati replied:
È possibile salvare nella P.IVA del partner il numero della P.IVA senza il prefisso, quindi possiamo fare sempre così. Quello che troviamo in `IdCodice` sarà scritto in `vat`

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+        if value := get_text(xml_tree, partner_section_xpath + "//Denominazione"):
+            vals["name"] = value
+        else:
+            vals["name"] = " ".join(
+                filter(
+                    None,
+                    [
+                        get_text(xml_tree, partner_section_xpath + "//Nome"),
+                        get_text(xml_tree, partner_section_xpath + "//Cognome"),
+                    ],
+                )
+            )
```
Possiamo usare invece `partner_firstname` come in `16.0`?

eLBati replied:
In importazione possiamo per ora tenerlo così. Installando `partner_firstname` dovrebbe comunque già fare lo split di `name`

monen17 replied:
Era già praticamente fatto in un altro punto, quindi l'ho aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            vals[field_name] = value
+
+        if province := get_text(xml_tree, partner_section_xpath + "//Provincia"):
+            if provinces := self.env["res.country.state"].search(
+                [("code", "=", province), ("country_id", "=", partner.country_id.id)]
+            ):
+                vals["state_id"] = fields.first(provinces).id
```
```suggestion
            if province := self.env["res.country.state"].search(
                [("code", "=", province), ("country_id", "=", partner.country_id.id)], limit=1
            ):
                vals["state_id"] = province.id
```

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+        if register_province := get_text(
+            xml_tree, partner_section_xpath + "//ProvinciaAlbo"
+        ):
+            if provinces := self.env["res.country.state"].search(
+                [
+                    ("code", "=", register_province),
+                    ("country_id", "=", partner.country_id.id),
+                ]
+            ):
+                vals["l10n_edi_it_register_province"] = fields.first(provinces).id
```
```suggestion
            if province := self.env["res.country.state"].search(
                [
                    ("code", "=", register_province),
                    ("country_id", "=", partner.country_id.id),
                ], limit=1
            ):
                vals["l10n_edi_it_register_province"] = province.id
```

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+        string="Do not update the contact from Electronic Invoice Details"
+    )
+    l10n_edi_it_register = fields.Char(string="Professional Register")
+    l10n_edi_it_register_province = fields.Many2one(
```
```suggestion
    l10n_edi_it_register_province_id = fields.Many2one(
```

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                vals["state_id"] = fields.first(provinces).id
+            else:
+                message = self.env._(
+                    f"Province ({province}) not present in your system"
```
Puoi verificare se questo viene tradotto correttamente?
Perché mi sa che la f-string viene valutata prima, andrebbe fatta la sostituzione con i parametri di `_`.

eLBati replied:
andrebbe modificato

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                vals["l10n_edi_it_register_province"] = fields.first(provinces).id
+            else:
+                message = self.env._(
+                    f"Register Province ({register_province}) not present in "
```
Puoi verificare se questo viene tradotto correttamente?
Perché mi sa che la f-string viene valutata prima, andrebbe fatta la sostituzione con i parametri di `_`.

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                    xml_tree, partner_info["section_xpath"], partner
+                )
+
+            if elements_stabile_organizzazione := xml_tree.xpath(
+                partner_info["section_xpath"] + "/StabileOrganizzazione"
+            ):
+                for element_stabile_organizzazione in elements_stabile_organizzazione:
+                    self.write(
+                        {
+                            "l10n_it_edi_stabile_organizzazione_indirizzo": get_text(
+                                element_stabile_organizzazione, ".//Indirizzo"
+                            ),
+                            "l10n_it_edi_stabile_organizzazione_civico": get_date(
+                                element_stabile_organizzazione, ".//NumeroCivico"
+                            ),
+                            "l10n_it_edi_stabile_organizzazione_cap": get_date(
+                                element_stabile_organizzazione, ".//CAP"
+                            ),
+                            "l10n_it_edi_stabile_organizzazione_comune": get_date(
+                                element_stabile_organizzazione, ".//Comune"
+                            ),
+                            "l10n_it_edi_stabile_organizzazione_provincia": get_date(
+                                element_stabile_organizzazione, ".//Provincia"
+                            ),
+                            "l10n_it_edi_stabile_organizzazione_nazione": get_date(
+                                element_stabile_organizzazione, ".//Nazione"
+                            ),
+                        }
+                    )
```
C'è solo un nodo `StabileOrganizzazione`, perché quindi un `for` loop? Tra l'altro se ci fossero più nodi i valori dei campi assegnati verrebbero sovrascritti.

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                "tax_amount": get_float(element, ".//AliquotaIVA"),
+                "wt_amount": get_text(element, ".//Ritenuta"),
+                "tax_kind": get_text(element, ".//Natura").upper(),
+                "admin_ref": get_text(element, ".//RiferimentoAmministrazione"),
```
Questo è già stato estratto sopra, possiamo usare quello?

monen17 replied:
:+1:

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            )
+            != 0
+        ):
+            error_message = self.env._(
+                f"Untaxed amount ({self.amount_untaxed}) "
+                f"does not match with "
+                f"e-invoice untaxed amount ({self.l10n_it_edi_amount_untaxed})"
+            )
```
Puoi verificare se questo viene tradotto correttamente?
Perché mi sa che la f-string viene valutata prima, andrebbe fatta la sostituzione con i parametri di `_`.

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                f"Taxed amount ({self.amount_tax}) "
+                f"does not match with "
+                f"e-invoice taxed amount ({self.l10n_it_edi_amount_tax})"
+            )
```
Puoi verificare se questo viene tradotto correttamente?
Perché mi sa che la f-string viene valutata prima, andrebbe fatta la sostituzione con i parametri di `_`.

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+                f"Total amount ({self.amount_total}) "
+                f"does not match with "
+                f"e-invoice total amount ({self.l10n_it_edi_amount_total})"
+            )
```
Puoi verificare se questo viene tradotto correttamente?
Perché mi sa che la f-string viene valutata prima, andrebbe fatta la sostituzione con i parametri di `_`.

eLBati replied:
per @monen17

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+            error_message = self.env._(
+                f"Untaxed amount ({self.amount_untaxed}) "
+                f"does not match with "
+                f"e-invoice untaxed amount ({self.l10n_it_edi_amount_untaxed})"
```
Citiamo anche l'arrotondamento visto che viene usato nel confronto?

eLBati replied:
possiamo aggiungere "con un arrotondamento di"

monen17 replied:
Aggiunto in https://github.com/Borruso/l10n-italy/pull/33

monen17 commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,874 @@
+        if (
+            self.l10n_it_edi_amount_untaxed
+            and float_compare(
+                self.amount_untaxed - self.l10n_it_edi_rounding,
+                abs(self.l10n_it_edi_amount_untaxed),
```
`l10n_it_edi_rounding` non andrebbe invece sommato a `l10n_it_edi_amount_untaxed` ?

eLBati replied:
sulla 16 è fatto così quindi ci fidiamo

monen17 replied:
:+1:

monen17 reviewed (changes requested):
Grazie delle modifiche!
Come detto stamattina in chiamata non ho guardato cosa c'è/manca rispetto a `16.0` tranne che per pochi punti, ho solo guardato come funziona cosa c'è nel README e se manca qualcosa.

micheledic commented:
@Borruso è stato mergiato il mio fix https://github.com/odoo/odoo/pull/217670 rilanciando i test non dovrebbero andare più in errore

monen17 commented:
@Borruso per risolvere i commenti da https://github.com/OCA/l10n-italy/pull/4411#discussion_r2215703663 in giù della revisione https://github.com/OCA/l10n-italy/pull/4411#pullrequestreview-3010673959 ho proposto https://github.com/Borruso/l10n-italy/pull/33.
Ora ha dei conflitti perché hai appena fatto diversi force-push, e vedo che i test falliscono.
Quando hai finito fammi sapere che faccio rebase di https://github.com/Borruso/l10n-italy/pull/33.
cc @eLBati

Borruso commented:
> @Borruso per risolvere i commenti da [#4411 (comment)](https://github.com/OCA/l10n-italy/pull/4411#discussion_r2215703663) in giù della revisione [#4411 (review)](https://github.com/OCA/l10n-italy/pull/4411#pullrequestreview-3010673959) ho proposto [Borruso#33](https://github.com/Borruso/l10n-italy/pull/33). Ora ha dei conflitti perché hai appena fatto diversi force-push, e vedo che i test falliscono. Quando hai finito fammi sapere che faccio rebase di [Borruso#33](https://github.com/Borruso/l10n-italy/pull/33). cc @eLBati

puoi fare rebuild 
potresti controllare l'errore del test su codice fiscale non corretto?

monen17 commented:
> > @Borruso per risolvere i commenti da [#4411 (comment)](https://github.com/OCA/l10n-italy/pull/4411#discussion_r2215703663) in giù della revisione [#4411 (review)](https://github.com/OCA/l10n-italy/pull/4411#pullrequestreview-3010673959) ho proposto [Borruso#33](https://github.com/Borruso/l10n-italy/pull/33). Ora ha dei conflitti perché hai appena fatto diversi force-push, e vedo che i test falliscono. Quando hai finito fammi sapere che faccio rebase di [Borruso#33](https://github.com/Borruso/l10n-italy/pull/33). cc @eLBati
> 
> puoi fare rebuild potresti controllare l'errore del test su codice fiscale non corretto?

L'errore è
> Invalid Codice Fiscale 'MRORSS90E25B111T': should be like 'MRTMTT91D08F205J' for physical person and '12345670546' for businesses.

<details><summary>Stack</summary>
<pre>
2025-07-18 15:39:23,847 268 ERROR odoo odoo.addons.l10n_it_edi_extension.tests.test_import_edi_extension_xml: ERROR: TestFatturaPAXMLValidation.test_import_zip
Traceback (most recent call last):
  File "/__w/l10n-italy/l10n-italy/l10n_it_edi_extension/tests/test_import_edi_extension_xml.py", line 155, in test_import_zip
    action = wizard_attachment_import.action_import()
  File "/__w/l10n-italy/l10n-italy/l10n_it_edi_extension/wizards/l10n_it_edi_import_file_wizard.py", line 88, in action_import
    move._l10n_it_edi_import_invoice(move, file_data, True)
  File "/__w/l10n-italy/l10n-italy/l10n_it_edi_extension/models/account_move.py", line 836, in _l10n_it_edi_import_invoice
    intermediary = self._l10n_it_edi_extension_create_partner(
  File "/__w/l10n-italy/l10n-italy/l10n_it_edi_extension/models/account_move.py", line 815, in _l10n_it_edi_extension_create_partner
    return self.env["res.partner"].create(partner_values)
  File "<decorator-gen-192>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 495, in _model_create_multi
    return create(self, [arg])
  File "/opt/odoo/addons/base_vat/models/res_partner.py", line 768, in create
    res = super().create(vals_list)
  File "<decorator-gen-194>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 496, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/addons/account/models/partner.py", line 824, in create
    return super().create(vals_list)
  File "<decorator-gen-97>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 496, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/addons/base/models/res_partner.py", line 806, in create
    partners = super().create(vals_list)
  File "<decorator-gen-140>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 496, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/addons/mail/models/mail_thread.py", line 274, in create
    threads = super(MailThread, self).create(vals_list)
  File "<decorator-gen-31>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 496, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/models.py", line 5017, in create
    records = self._create(data_list)
  File "/opt/odoo/odoo/models.py", line 5275, in _create
    records._validate_fields(name for data in data_list for name in data['stored'])
  File "/opt/odoo/odoo/models.py", line 1631, in _validate_fields
    check(self)
  File "/__w/l10n-italy/l10n-italy/l10n_it_edi_extension/models/res_partner.py", line 42, in validate_codice_fiscale
    res = super().validate_codice_fiscale()
  File "/opt/odoo/addons/l10n_it_edi/models/res_partner.py", line 159, in validate_codice_fiscale
    raise UserError(_("Invalid Codice Fiscale '%s': should be like 'MRTMTT91D08F205J' for physical person and '12[345](https://github.com/OCA/l10n-italy/actions/runs/16374490895/job/46270888764?pr=4411#step:8:346)670546' for businesses.", record.l10n_it_codice_fiscale))
odoo.exceptions.UserError: Invalid Codice Fiscale 'MRORSS90E25B111T': should be like 'MRTMTT91D08F205J' for physical person and '12345670546' for businesses.
</pre>
</details> 
(da https://github.com/OCA/l10n-italy/actions/runs/16374490895/job/46270888764?pr=4411#step:8:327)

Ho verificato che il codice fiscale indicato effettivamente non è corretto in https://telematici.agenziaentrate.gov.it/VerificaCF/IVerificaCf.jsp:
<img width="558" height="180" alt="image" src="https://github.com/user-attachments/assets/69f2f280-7829-4792-a343-48954b085635" />

Questo succede perché esiste un terzo intermediario nello zip dei test che ha quel codice fiscale, e fallisce la verifica.
Aggiungo la correzione in https://github.com/Borruso/l10n-italy/pull/33.

monen17 reviewed (changes requested):
Grazie delle modifiche!
Come d'accordo parte delle correzioni le ho fatte in https://github.com/Borruso/l10n-italy/pull/33.

Tra le mie modifiche e le tue credo rimangano aperti solo:
- https://github.com/OCA/l10n-italy/pull/4411#discussion_r2200994569
- https://github.com/OCA/l10n-italy/pull/4411#discussion_r2201406138

Ho notato che in https://github.com/OCA/l10n-italy/pull/4411/commits/45cdc78ba8042b147ba89d9bc840171669d4318f sono finite diverse modifiche che non erano nel commit originale (https://github.com/Borruso/l10n-italy/pull/23/commits/50127b6fdfd6957a63bcb31913e38106dd066f9b), puoi ripristinare il commit originale e mettere le nuove modifiche in un altro commit?
Non posso farlo in una PR perché creerebbe conflitti con https://github.com/OCA/l10n-italy/pull/4411/commits/45cdc78ba8042b147ba89d9bc840171669d4318f e non si può fare dopo il merge.

eLBati commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,158 @@
+2. Valorizza i seguenti campi/nodi della fattura elettronica esportata:
+
+   - `<RiferimentoAmministrazione>` (sia sulla riga o nell'azienda): identificativo utilizzato per uso amministrativo/gestionale interno. È un campo libero che può essere utilizzato per inserire riferimenti specifici richiesti dalla Pubblica Amministrazione o altri riferimenti utili per la gestione amministrativa.
+   - `<StabileOrganizzazione>`: rappresenta i dati della sede operativa stabile del cedente/prestatore in Italia se diversa dalla sede legale
```
```suggestion
   - `<StabileOrganizzazione>`: rappresenta i dati della sede operativa stabile del cedente/prestatore in Italia se diversa dalla sede legale. Da impostare in odoo nell'azienda, tab "informazioni generali"
```

Borruso replied:
fatto

eLBati commented on l10n_it_edi_extension/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,158 @@
+   - `<RiferimentoAmministrazione>` (sia sulla riga o nell'azienda): identificativo utilizzato per uso amministrativo/gestionale interno. È un campo libero che può essere utilizzato per inserire riferimenti specifici richiesti dalla Pubblica Amministrazione o altri riferimenti utili per la gestione amministrativa.
+   - `<StabileOrganizzazione>`: rappresenta i dati della sede operativa stabile del cedente/prestatore in Italia se diversa dalla sede legale
+   - `<Causale>`: in questo caso non c'è un campo apposito, ma trascrive i "Termini e condizioni" della fattura
+   - `<Art73>`: indica se il documento è stato emesso secondo modalità e termini stabiliti con decreto ministeriale ai sensi dell'articolo 73 del DPR 633/72
```
```suggestion
   - `<Art73>`: indica se il documento è stato emesso secondo modalità e termini stabiliti con decreto ministeriale ai sensi dell'articolo 73 del DPR 633/72. Da impostare in odoo nell'azienda o direttamente in fattura.
```

Borruso replied:
fatto

Commit:
[IMP]l10n_it_edi_extension: flag for create partner

Borruso committed:
[REF] l10n_it_edi_extension: refactor tax representative and intermediary creation

monen17 commented:
/ocabot migration l10n_it_fatturapa

monen17 commented:
/ocabot migration l10n_it_fatturapa_in

monen17 commented:
/ocabot migration l10n_it_fatturapa_out

monen17 commented:
/ocabot migration l10n_it_fiscal_payment_term

monen17 commented:
/ocabot migration l10n_it_fiscalcode

monen17 commented:
/ocabot migration l10n_it_ipa

monen17 commented:
/ocabot migration l10n_it_pec

monen17 commented:
/ocabot migration l10n_it_rea

monen17 commented:
/ocabot migration l10n_it_vat_payability

eLBati commented on l10n_it_edi_extension/models/account_move.py:
```diff
@@ -0,0 +1,836 @@
+            partner = self._l10n_it_edi_create_partner(
+                xml_tree, partner_info["section_xpath"], vat, codice_fiscale
+            )
+            if not partner.l10n_edi_it_electronic_invoice_no_contact_update:
```
Questo inoltre è da spostare in `_l10n_it_edi_import_invoice`

Borruso replied:
fatto

Borruso committed:
[REF] l10n_it_edi_extension: refactor update partner from FatturaPA attachment

monen17 commented:
Puoi mergiare https://github.com/Borruso/l10n-italy/pull/33?

monen17 commented:
Porto avanti con https://github.com/OCA/l10n-italy/pull/4858.
